README.md
### Description

Source code for my [YATM](https://chrome.google.com/webstore/detail/yatm-yet-another-tab-mana/cleopomnamfenlfpglofnooemoakdgef?hl=en) ("Yet Another Tab Manager") Chrome tab manager extension

### Technology Stack
- Language: [TypeScript](https://www.typescriptlang.org)
- Frontend: HTML

### Screenshots

![screen1](assets/screenshots/main.png)
![screen1](assets/screenshots/search.png)


assets/history.txt
v2.3
- Moved to Typescript

v2.2
- fixed a bug that was preventing YATM from switching from "popup" to "side panel" mode outside of dev mode

v2.1
- Added buttons to access the configuration dialog & the help video
- Added ability to switch between popup mode (the default) or side panel mode
- Fixed a nasty bug that was causing the side panel to randomly close when previewing tabs
- Removed the capacity to close YATM when clicking anywhere in the main browser window (this was buggy anyways)

v2.0
- YATM now opens as a side panel:
-- click on a tab in YATM to preview it
-- double click on a tab in YATM to close YATM
-- click anywhere in the main browser window to close YATM
- Side panel width is adjustable using your mouse
- Removed sort button

v1.5
- Renamed extension from "Tab List" to "YATM"

v1.4
- Add search result highlights
- Fix focus issue on search input element
- Renamed "sort by domain" to "sort by website"

v1.3
- add "X" icon in search box
- fix bug in sorting on tab name

v1.2
- initial release


src/js/utils.ts
const IS_LOG_TO_LOCALSTORAGE = true;
const DEFAULT_PANEL_WIDTH = 500;
const DEFAULT_IS_POPUP = true;
const PANEL_WIDTH_STORAGE_KEY = "panelWidth";
const IS_POPUP_STORAGE_KEY = "isPopup";
const LOGS_STORAGE_KEY = "logs";

function log(log:string):void {
    let now:Date = new Date();
    log = now.getHours() + ":" + now.getMinutes() + ":" + now.getSeconds() + " " + log;
    console.log(log);
    if (IS_LOG_TO_LOCALSTORAGE) {
        let logs:string = localStorage.getItem(LOGS_STORAGE_KEY);
        if (!logs) logs = "";
        logs += (log + "\n");
        localStorage.setItem(LOGS_STORAGE_KEY, logs);
    }
}

function saveIsPopupToStorage(isPopup:boolean):void {
    log(">> utils.saveIsPopupToStorage: isPopup=" + isPopup);
    localStorage.setItem(IS_POPUP_STORAGE_KEY, isPopup.toString());
    log("<< utils.saveIsPopupToStorage");
}

function getIsPopupFromStorage():boolean {
    log(">> utils.getIsPopupFromStorage");
    // https://www.w3schools.com/html/html5_webstorage.asp
    let isPopupAsString:string = localStorage.getItem(IS_POPUP_STORAGE_KEY);
    let isPopup:boolean = isPopupAsString ? isPopupAsString === 'true' : DEFAULT_IS_POPUP;
    log("<< utils.getIsPopupFromStorage: isPopup=" + isPopup);
    return isPopup;
}

function savePanelWidthToStorage(panelWidth:number):void {
    log(">> utils.savePanelWidthToStorage: panelWidth=" + panelWidth);
    let panelWidthAsString:string = panelWidth != null ? panelWidth.toString() : "";
    localStorage.setItem(PANEL_WIDTH_STORAGE_KEY, panelWidthAsString);
    log("<< utils.savePanelWidthToStorage");
}

function getPanelWidthFromStorage():number {
    log(">> utils.getPanelWidthFromStorage");
    // https://www.w3schools.com/html/html5_webstorage.asp
    // I could have used https://developer.chrome.com/extensions/storage but those methods are asynchronous and I don't feel like using them
    // localStorage.getItem() returns 'null' if there is no key (which will be the case if YATM was just installed)
    // note that uninstalling YATM removes the localStorage
    let panelWidthAsString:string = localStorage.getItem(PANEL_WIDTH_STORAGE_KEY);
    let panelWidth:number = panelWidthAsString ? parseInt(panelWidthAsString) : DEFAULT_PANEL_WIDTH;
    log("<< utils.getPanelWidthFromStorage: panelWidth=" + panelWidth);
    return panelWidth;
}

/*
 * https://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript
 */
function getQueryParameterByName(name):string {
    let url:string = window.location.href;
    //log(">> utils.getQueryParameterByName: url=" + url + ", name=" + name);
    name = name.replace(/[\[\]]/g, "\\$&");
    let regex = new RegExp("[?&]" + name + "(=([^&#]*)|&|#|$)"),
        results = regex.exec(url);
    let result;
    if (!results) {
        result = null;
    } else if (!results[2]) {
        result = '';
    } else {
        result = decodeURIComponent(results[2].replace(/\+/g, " "));
    }
    //log("<< utils.getQueryParameterByName: result=" + result);
    return result;
}

function byid(idValue):HTMLElement {
    return document.getElementById(idValue);
}

function byclass(elt, className):HTMLElement {
    return elt.getElementsByClassName(className)[0];
}

function newElt(tagName:string, attributes, textContent?:string):HTMLElement {
    //log(">> newElt= tagName=" + tagName + ", attributes=" + JSON.stringify(attributes) + ", textContent=" + textContent);
    let elt:HTMLElement = document.createElement(tagName);
    for (let key in attributes) {
        elt.setAttribute(key, attributes[key]);
    }
    if (textContent !== undefined) {
        elt.textContent = textContent;
    }
    //log("<< newElt=" + elt.outerHTML);
    return elt;
}

function addTextHighlight(elt:HTMLElement, text:string, highlightClass:string):void {
    //log(">> addTextHighlight: elt='" + elt.outerHTML + "', text='" + text + "', class='" + highlightClass + "'");
    let highlightedText:string = "<span class=\"" + highlightClass + "\">" + text + "</span>";
    elt.innerHTML = elt.innerHTML.replace(new RegExp(text, "i"), highlightedText); // "g"=global "i"=case insensitive
    //log("<< addTextHighlight: elt='" + elt.outerHTML + "'");
}

function removeTextHighlight(elt:HTMLElement):void {
    //log(">> removeTextHighlight: elt='" + elt.outerHTML + "'");
    elt.innerHTML = elt.textContent;
    //log("<< removeTextHighlight: elt=" + elt.outerHTML + "'");
}

src/js/options.ts
// http://blogs.microsoft.co.il/gilf/2013/07/22/quick-tip-typescript-declare-keyword/
declare var chrome;

function saveOptions() {
    let panelWidthAsString:string = (<HTMLInputElement>byid('panelWidth')).value;
    let panelWidth:number = panelWidthAsString ? parseInt(panelWidthAsString) : null;
    panelWidth = panelWidth < 200 ? 200 : panelWidth; // basic validation

    let displayModeElt = <HTMLSelectElement>byid("displayMode");
    let displayMode:string = displayModeElt.options[displayModeElt.selectedIndex].value;
    let isPopup:boolean = (displayMode === "popup");

    log(">> options.saveOptions: panelWidth=" + panelWidth + ", isPopup=" + isPopup);
    savePanelWidthToStorage(panelWidth);
    let switchDisplayMode:boolean = getIsPopupFromStorage() !== isPopup;
    saveIsPopupToStorage(isPopup);
    if (switchDisplayMode) {
        // reloading the whole extension to change the way the side panel is displayed
        // the issue of doing this here is that this going to dismiss the options dialog
        log("options.saveOptions: reloading extension");
        // - in dev mode, calling runtime.reload() loads the background script
        // - in production mode, calling runtime.reload() does NOT load the background script
        chrome.runtime.getBackgroundPage(function (bg) {
            // note that this will invoked init() twice in case the background page is already loaded
            // that's OK since init() is clever init not to do the same things twice
            bg.init();
            log("options.saveOptions: done reloading extension");
        });
    }
    let status:HTMLElement = byid('status');
    status.textContent = ' Options saved!';
    setTimeout(function () {
        status.textContent = '';
    }, 750);
    log("<< options.saveOptions");
}

function restoreOptions() {
    log(">> options.restoreOptions");
    (<HTMLInputElement>byid('panelWidth')).value = getPanelWidthFromStorage().toString();
    if (getIsPopupFromStorage()) {
        (<HTMLInputElement>byid('displayMode')).value = "popup";
    } else {
        (<HTMLInputElement>byid('displayMode')).value = "sidePanel";
    }
    log("<< options.restoreOptions");
}

document.addEventListener('DOMContentLoaded', restoreOptions);
document.getElementById('save').addEventListener('click', saveOptions);

src/js/background.ts
// http://blogs.microsoft.co.il/gilf/2013/07/22/quick-tip-typescript-declare-keyword/
declare var chrome;
let g_wasMainWindowShrinked;
let g_originalMainWindowWidth;
let g_mainWindowId;
let g_panelWindowId;
let g_clickedTabId;

function getGlobalsAsString():string {
    let result:string = "";
    result += "g_wasMainWindowShrinked=" + g_wasMainWindowShrinked + ", ";
    result += "g_originalMainWindowWidth=" + g_originalMainWindowWidth + ", ";
    result += "g_mainWindowId=" + g_mainWindowId + ", ";
    result += "g_panelWindowId=" + g_panelWindowId + ", ";
    result += "g_clickedTabId=" + g_clickedTabId + ", ";
    return result;
}

/*
 * We potentially need to resize the main window size to leave room for the panel
 *
 * availableWidth = spacing between the right side of the window and the right side of the screen
 * availableWidth = 150 & panelWidth=100 -> delta = 50 -> delta is >= 0 so no need to reduce main window size
 * availableWidth = 20  & panelWidth=100 -> delta = -80 -> delta is <0 so need to reduce window size
 */
function getMainWindowNewWidth(mainWindow):number {
    log(">> bg.getMainWindowNewWidth");
    let availablePanelWidth:number = screen.width - (mainWindow.left + mainWindow.width);
    log("bg.getMainWindowNewWidth: screen.width=" + screen.width + ", window.left=" + mainWindow.left + ", window.width=" + mainWindow.width + ' -> availablePanelWidth=' + availablePanelWidth);
    // TODO: need to check the scenario where for some reason the savedPanelWidth is way too big
    // (larger than the spacing between the MINIMUM left side of the main window and the right side of the screen
    let panelWidth:number = getPanelWidthFromStorage();
    let delta:number = availablePanelWidth - panelWidth;
    log("bg.getMainWindowNewWidth: availablePanelWidth=" + availablePanelWidth + ", panelWidth=" + panelWidth + ' -> delta=' + delta);
    let mainWindowNewWidth:number;
    if (delta >= 0) {
        mainWindowNewWidth = mainWindow.width;
        g_wasMainWindowShrinked = false;
    } else {
        mainWindowNewWidth = mainWindow.width + delta;
        g_wasMainWindowShrinked = true;
        g_originalMainWindowWidth = mainWindow.width;
    }
    if (mainWindowNewWidth < 0) {
        // crap: not enough size, what should we do?
    }
    log("<< bg.getMainWindowNewWidth: mainWindowNewWidth=" + mainWindowNewWidth);
    return mainWindowNewWidth;
}


function initSidePanel():void {
    log(">> bg.initSidePanel");
    chrome.windows.onRemoved.addListener(onWindowsRemovedListener);
    chrome.windows.getCurrent({}, function (mainWindow) {
        g_mainWindowId = mainWindow.id;
        let mainWindowNewWidth:number = getMainWindowNewWidth(mainWindow);
        // optionally resize main window to leave room for right panel:
        chrome.windows.update(
            mainWindow.id,
            {'width': mainWindowNewWidth},
            function (updatedWindow) {
                // close previous panel, if any (in case the user opens the extension from a new window)
                if (g_panelWindowId) { // required since windows.remove() fails if the ID is undefined (checking for lastError won't work)
                    chrome.windows.remove(g_panelWindowId, function () {
                        if (chrome.runtime.lastError) {
                            // that's OK, it just means that there is no other panel window
                        }
                    });
                }
                // create side panel:
                chrome.windows.create({
                    'url': chrome.extension.getURL('src/html/panel.html?mainWindowId=' + g_mainWindowId + "&isPopup=false"),
                    'left': updatedWindow.left + updatedWindow.width,
                    'top': updatedWindow.top,
                    'width': getPanelWidthFromStorage(),
                    'height': screen.height,
                    'type': 'popup'
                }, function (panelWindow) {
                    g_panelWindowId = panelWindow.id;
                    log("bg.initSidePanel: g_mainWindowId=" + g_mainWindowId + ", g_panelWindowId=" + g_panelWindowId);
                    chrome.tabs.onActivated.addListener(onActivatedListener);
                });
            }
        );
    });
    log("<< bg.initSidePanel");
}


function onActivatedListener(activeInfo):void {
    log(">> bg.onActivatedListener: activeWindowId=" + activeInfo.windowId + ", activeTabId=" + activeInfo.tabId + ", g_clickedTabId=" + g_clickedTabId);
    if (g_clickedTabId) {
        // for some unknown reason, when using chrome.tabs.update to make a tab active, sometimes the side panel keeps the focus and sometimes loses it
        // to handle the second scenario, we always give back the focus to the side panel after a tab was selected
        chrome.windows.update(g_panelWindowId, {focused: true}, function () {
            g_clickedTabId = null;
        })
    }
    log("<< bg.onActivatedListener");
}


function restoreMainWindowSize():void {
    log(">> bg.restoreMainWindowSize: g_mainWindowId=" + g_mainWindowId + ", g_wasMainWindowShrinked=" + g_wasMainWindowShrinked);
    if (g_mainWindowId && g_wasMainWindowShrinked) {
        chrome.windows.get(g_mainWindowId, {}, function (mainWindow) {
            if (!chrome.runtime.lastError) {
                log("bg.restoreMainWindowSize: restoring size of window with id=" + mainWindow.id);
                chrome.windows.update(
                    mainWindow.id,
                    {'width': g_originalMainWindowWidth}
                );
            }
        });
    }
    log("<< bg.restoreMainWindowSize");
}


function onWindowsRemovedListener(removedWindowID):void {
    log(">> bg.onWindowsRemovedListener: removedWindowID=" + removedWindowID);
    if (removedWindowID === g_panelWindowId) { // panel was closed
        restoreMainWindowSize();
        chrome.windows.onActivated.removeListener(onActivatedListener);
        chrome.windows.onRemoved.removeListener(onWindowsRemovedListener);
        g_panelWindowId = null;
        g_mainWindowId = null;
    } else if ((removedWindowID === g_mainWindowId) && g_panelWindowId) { // main window was closed
        log("bg.onWindowsRemovedListener: removing g_panelWindowId=" + g_panelWindowId);
        // this is going to fire the windows.onRemoved listener on the panel window which will take care of the cleanup:
        chrome.windows.remove(g_panelWindowId);
    }
    log("<< bg.onWindowsRemovedListener");
}

// @ts-ignore
function init():void {
    log(">> bg.init()");
    if (getIsPopupFromStorage()) {
        chrome.browserAction.getPopup({}, function (popup) {
            if (!popup) {
                chrome.browserAction.setPopup({popup: "src/html/panel.html?isPopup=true"});
            }
        });
    } else {
        // browserAction.onClicked() will NOT fire if the browser action has a popup so we first remove it
        chrome.browserAction.setPopup({popup: ""});
        let hasListener = chrome.browserAction.onClicked.hasListener(initSidePanel);
        if (!hasListener) {
            chrome.browserAction.onClicked.addListener(initSidePanel);
        }
    }
    log("<< bg.init()");
}

// Because the listeners themselves only exist in the context of the event page, you must use addListener each time the event page loads
init();

/*
 * Official documentation: Fired when a **profile** that has this extension installed first starts up
 *
 * in devmode:
 * - NOT invoked when you install the unpacked extension
 */
chrome.runtime.onStartup.addListener(function () {
    log(">> << bg.runtime.onStartup");
});

/*
 * Official documentation: Fired when the extension is first installed, when the extension is updated to a new version, and when Chrome is updated to a new version.
 *
 * in dev mode:
 * - invoked when you install the unpacked extension
 * - invoked when you click on the "reload" button on the extension page OR when you reload the extension page
 * - NOT invoked when you click on the "inspect views: background page (inactive)" link (however main() is invoked)
 * - NOT invoked when you click on the "inspect views: background page" link since the background page is already loaded (main() is also NOT invoked)
 */
chrome.runtime.onInstalled.addListener(function () {
    log(">> << bg.runtime.onInstalled");
});

/*
 * Official documentation: Sent to the event page just before it is unloaded
 *
 * in dev mode:
 * - invoked when the background page becomes inactive
 */
chrome.runtime.onSuspend.addListener(function () {
    log(">> << bg.runtime.onSuspend");
});

chrome.runtime.onSuspendCanceled.addListener(function () {
    log(">> << bg.runtime.onSuspendCanceled");
});


src/js/panel.ts
// http://blogs.microsoft.co.il/gilf/2013/07/22/quick-tip-typescript-declare-keyword/
declare var chrome;
declare var Sortable;
let g_sortable;
let g_isPopup:boolean;

/*
 *
 */
// @ts-ignore
function init():void {
    let g_mainWindowId:number = parseInt(getQueryParameterByName("mainWindowId"));
    g_isPopup = "true" === getQueryParameterByName("isPopup");
    log(">> panel.init: mainWindowId=" + g_mainWindowId + ", g_isPopup=" + g_isPopup);

    if (g_isPopup) {
        // extension is displayed as a standard popup
        // popup dynamically adjusts to its content: need to explicitly set the width in case the search returns nothing
        byid("main").setAttribute("style", "width:" + getPanelWidthFromStorage() + "px; height: 100px;");

    } else {
        // extension is displayed in a side panel
        window.addEventListener("resize", onWindowResized);
    }

    g_sortable = createSortable(byid("list"));

    let queryParams = g_mainWindowId ? {windowId: g_mainWindowId} : {currentWindow: true};
    chrome.tabs.query(queryParams, createTabList);
    /////////
    // SEARCH
    /////////
    byid("searchInput").addEventListener("keyup", search);
    // https://www.w3schools.com/jsref/event_onsearch.asp
    byid("searchInput").addEventListener("search", search);
    byid("searchInput").focus();
    ///////////
    // Options
    //////////
    byid("config").addEventListener("click", function () {
        chrome.runtime.openOptionsPage();
    });
    ///////////
    // Help
    //////////
    byid("help").addEventListener("click", function () {
        window.open("https://youtu.be/S5v1gYUEqmk");
        //chrome.runtime.getBackgroundPage(function() {});
    });
    /////////
    // RENAME
    /////////
    byid("renameTabNameInput").addEventListener("keyup", function (event) {
        event.preventDefault();
        if (event.keyCode === 13) {
            byid("renameOkButton").click();
        }
    });
    byid("renameCancelButton").addEventListener("click", function () {
        byid("renameDialog").style.display = "none";
    });
    byid("renameOkButton").addEventListener("click", rename);
    log("<< panel.init");
}


function onWindowResized(event):void {
    let width:number = event.currentTarget.outerWidth;
    log(">> panel.onWindowResized:width=" + width);
    savePanelWidthToStorage(width);
    log("<< panel.onWindowResized");
}


function search(event):void {
    let searched:string = event.target.value;
    log(">> panel.search: searched='" + searched + "'");
    if (searched.length > 0) {
        g_sortable.option("disabled", true);
        let liElts:NodeList = byid('list').childNodes;
        for (let i = 0; i < liElts.length; i++) {
            let nextLiElt:HTMLLIElement = <HTMLLIElement>liElts[i];
            let titleElt:HTMLSpanElement = byclass(nextLiElt, "title");
            let matches = titleElt.textContent.match(new RegExp(searched, "i"));
            if (matches) {
                nextLiElt.style.display = "inline-block";
                byclass(nextLiElt, "drag").style.cursor = "no-drop";
                byclass(nextLiElt, "drag").title = "drag'n'drop is disabled when searching";
                removeTextHighlight(titleElt);
                addTextHighlight(titleElt, matches[0], "searched");
            } else {
                nextLiElt.style.display = "none";
            }
        }
    } else {
        g_sortable.option("disabled", false);
        location.reload();
    }
    log("<< panel.search");
}


function rename():void {
    log(">> panel.rename");
    let tabId:string = (<HTMLInputElement>byid('renameTabIdInput')).value;
    let tabName:string = (<HTMLInputElement>byid('renameTabNameInput')).value;
    chrome.tabs.executeScript(parseInt(tabId), {code: "document.title = '" + tabName + "'"}, function (results) {
        byid("renameDialog").style.display = "none";
        byclass(byid(tabId), "title").textContent = tabName;
    });
    log("<< panel.rename");
}


function previewTab(event):void {
    let tabId:number = parseInt(event.currentTarget.getAttribute("id"));
    log(">> panel.previewTab: tabId=" + tabId);
    chrome.runtime.getBackgroundPage(function (bgPage) {
        bgPage.g_clickedTabId = tabId;
        chrome.tabs.update(tabId, {active: true});
    });
    let liElts:NodeList = byid('list').childNodes;
    for (let i = 0; i < liElts.length; i++) {
        // https://www.w3schools.com/jsref/prop_element_classlist.asp
        // TODO: do not loop over all the nodes: find the ones who have the "selected" class and remove it
        (<HTMLLIElement>liElts[i]).classList.remove("selected");
    }
    // https://developer.mozilla.org/en-US/docs/Web/API/Event
    // event.target actually points to the HTML element you clicked on it and NOT to the "li" element:
    event.currentTarget.classList.add("selected");
    log("<< panel.previewTab");
}


function openTab(event):void {
    log(">> panel.openTab");
    let tabId:string = event.currentTarget.getAttribute("id");
    chrome.tabs.update(parseInt(tabId), {active: true});
    window.close();
    log(">> panel.openTab");
}


function closeTab(event):void {
    log(">> panel.closeTab");
    let liElt:HTMLLIElement = event.currentTarget.parentNode;
    let isSelected:boolean = liElt.classList.contains("selected");
    let tabId:string = liElt.getAttribute("id");
    log("panel.closeTab: tabId=" + tabId + ", isSelected=" + isSelected);
    chrome.tabs.remove(parseInt(tabId), function () {
        liElt.parentNode.removeChild(liElt);
        // we've closed the currently selected tab -> need to find which one is now selected
        if (isSelected) {
            // required so that the newly selected tab is properly highlighted in the list
            // for some reason, invoking "chrome.tabs.query({active:true})" returns the
            // tab that was just closed by chrome.tabs.remove() so we can't use it.
            location.reload();
        }
    });
    event.stopPropagation();
    event.preventDefault();
    log("<< panel.closeTab");
}


function showRenameDialog(event):void {
    log(">> panel.showRenameDialog");
    (<HTMLInputElement>byid('renameTabIdInput')).value = event.currentTarget.parentNode.getAttribute("id");
    let renameTabNameInputElt:HTMLInputElement = <HTMLInputElement>byid('renameTabNameInput');
    renameTabNameInputElt.value = byclass(event.currentTarget.parentNode, "title").innerHTML;
    renameTabNameInputElt.setSelectionRange(0, 0);

    let dialogElt:HTMLElement = byid('renameDialog');
    dialogElt.style.display = "inline-block";
    // https://stackoverflow.com/questions/15615552/get-div-height-with-plain-javascript
    dialogElt.style.top = event.clientY + dialogElt.clientHeight + "px"; // event.clientY is the mouse position
    renameTabNameInputElt.focus();

    event.stopPropagation();
    event.preventDefault();
    log("<< panel.showRenameDialog");
}


function createTabList(tabs):void {
    log(">> panel.createTabList: nbTabs=" + tabs.length);
    let mainElt:HTMLElement = byid('main');
    let ulElt:HTMLElement = byid('list');

    for (let i = 0; i < tabs.length; i++) {
        let nextTab = tabs[i];

        // https://developer.mozilla.org/en-US/docs/Web/API/Element
        let liElt:HTMLElement = newElt("li", {id: nextTab.id, class: "item" + (nextTab.selected ? ' selected' : '')});
        if (g_isPopup) {
            liElt.addEventListener('click', openTab);
        } else {
            liElt.addEventListener('click', previewTab);
            liElt.addEventListener('dblclick', openTab);
        }
        ulElt.appendChild(liElt);
        let dragElt:HTMLElement = newElt("span", {class: "drag", title: "Drag'n'drop to reorder tabs"});
        liElt.appendChild(dragElt);
        let faviconElt:HTMLElement = newElt("img", {
            class: "favicon",
            alt: "",
            src: (nextTab.favIconUrl ? nextTab.favIconUrl : '../png/empty.png')
        });
        liElt.appendChild(faviconElt);
        let spanTitle:string;
        if (g_isPopup) {
            spanTitle = "Click to open tab";
        } else {
            spanTitle = "Click to preview this tab, Double click to open it";
        }
        let titleElt:HTMLElement = newElt("span", {
            class: "title",
            title: spanTitle
        }, nextTab.title);
        liElt.appendChild(titleElt);
        let editElt:HTMLElement = newElt("span", {class: "edit", title: "Click to rename this tab"});
        editElt.addEventListener('click', showRenameDialog);
        liElt.appendChild(editElt);
        let closeElt:HTMLElement = newElt("span", {class: "close", title: "Click to close this tab"});
        closeElt.addEventListener('click', closeTab);
        liElt.appendChild(closeElt);

    }

    mainElt.appendChild(ulElt);
    log("<< panel.createTabList");

}

/*
 $ https://github.com/RubaXa/Sortable
 init exemple here: https://github.com/RubaXa/Sortable/wiki/Sorting-with-the-help-of-HTML5-Drag'n'Drop-API
 */
function createSortable(elt):void {
    log(">> panel.createSortable");
    let sortable = Sortable.create(elt, {
            animation: 200,
            onStart: function (event) {
                byclass(event.item, "favicon").style.display = "none";
                byclass(event.item, "drag").style.display = "none";
            },
            onEnd: function (event) {
                // now we restore what was hidden in onStart()
                // TODO: this is really a hack since we assume that these 2 elements are always visible
                byclass(event.item, "favicon").style.display = "inline-block";
                byclass(event.item, "drag").style.display = "inline-block";

                let tabId:string = event.item.getAttribute("id");
                //log("createSortable.onEnd: liEl=" + liEl + ", oldIndex=" + event.oldIndex, ", newIndex=" + event.newIndex);
                chrome.tabs.move(parseInt(tabId), {index: event.newIndex}, function (movedTab) {
                    // note that the movedTab.index property may NOT be the actual one: if you try to drop the element
                    // between 2 pinned tabs, Chrome will automatically move the tab AFTER them but tab.index will
                    // not reflect that. We need to invoke tabs.get() to know where the tab was actually inserted
                    chrome.tabs.get(parseInt(tabId), function (tab) {
                        // https://developer.mozilla.org/en-US/docs/Web/API/Location/reload
                        if (movedTab.index !== tab.index) location.reload();
                    })

                });
            }
        }
    );
    log("<< panel.createSortable");
    return sortable;

}


document.addEventListener("DOMContentLoaded", init);
